File: services/reports-service/src/ReportsService.Api/Security/RedisPermissionResolver.cs
namespace ReportsService.Api.Security;

public sealed class RedisPermissionResolver : IPermissionResolver
{
    private readonly IRoleAccessGateway _roleAccessGateway;

    public RedisPermissionResolver(IRoleAccessGateway roleAccessGateway)
    {
        _roleAccessGateway = roleAccessGateway;
    }

    public async Task<PermissionDecision> ResolveAsync(
        IReadOnlyCollection<string> roles,
        string permission,
        string resource,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(roles);
        ArgumentException.ThrowIfNullOrWhiteSpace(permission);
        ArgumentException.ThrowIfNullOrWhiteSpace(resource);

        if (roles.Count == 0)
        {
            return new PermissionDecision(false, Array.Empty<string>(), Array.Empty<string>(), "Authenticated token does not contain a valid role");
        }

        var matchedRoles = new List<string>();
        var matchedPermissions = new List<string>();

        foreach (var role in roles)
        {
            var record = await _roleAccessGateway.GetRoleAccessAsync(role, cancellationToken);
            if (record is null)
            {
                continue;
            }

            if (record.Permissions.Contains(permission, StringComparer.Ordinal))
            {
                matchedRoles.Add(record.Role);
                matchedPermissions.Add(permission);
            }
        }

        if (matchedPermissions.Count > 0)
        {
            return new PermissionDecision(true, matchedRoles.Distinct(StringComparer.Ordinal).ToArray(), matchedPermissions.Distinct(StringComparer.Ordinal).ToArray(), "Permission granted by effective role access");
        }

        return new PermissionDecision(false, matchedRoles.Distinct(StringComparer.Ordinal).ToArray(), Array.Empty<string>(), $"Effective permission not granted for resource {resource}");
    }
}

File: services/orders-service/src/main/java/com/zcorp/ordersservice/security/RedisPermissionResolver.java
package com.zcorp.ordersservice.security;

import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;

import org.springframework.stereotype.Component;

@Component
public class RedisPermissionResolver implements PermissionResolver {

    private final RoleAccessGateway roleAccessGateway;

    public RedisPermissionResolver(RoleAccessGateway roleAccessGateway) {
        this.roleAccessGateway = Objects.requireNonNull(roleAccessGateway, "roleAccessGateway must not be null");
    }

    @Override
    public PermissionDecision resolve(Set<String> roles, String permission, String resource) {
        Objects.requireNonNull(roles, "roles must not be null");
        Objects.requireNonNull(permission, "permission must not be null");
        Objects.requireNonNull(resource, "resource must not be null");

        if (roles.isEmpty()) {
            return new PermissionDecision(false, Set.of(), Set.of(), "Authenticated token does not contain a valid role");
        }

        Set<String> matchedRoles = new LinkedHashSet<>();
        Set<String> matchedPermissions = new LinkedHashSet<>();

        for (String role : roles) {
            roleAccessGateway.getRoleAccess(role)
                    .ifPresent(record -> {
                        if (record.permissions().contains(permission)) {
                            matchedRoles.add(record.role());
                            matchedPermissions.add(permission);
                        }
                    });
        }

        if (matchedPermissions.contains(permission)) {
            return new PermissionDecision(true, Set.copyOf(matchedRoles), Set.copyOf(matchedPermissions), "Permission granted by effective role access");
        }

        return new PermissionDecision(false, Set.copyOf(matchedRoles), Set.of(), "Effective permission not granted for resource " + resource);
    }
}